### PR TITLE
feat: add replication over tls

### PR DESCRIPTION
**Pull request by kostasrim**
1. Introduces `tls_replication` flag to allow tls connections for replicas
2. Depends on helio PR https://github.com/romange/helio/pull/98
3. Addresses #1418

### CODE REVIEW

File: src/server/replica.cc [hunk 1]
```diff
@@ -129,11 +140,49 @@ vector<vector<unsigned>> Partition(unsigned num_flows) {
 std::string Replica::MasterContext::Description() const {
   return absl::StrCat(host, ":", port);
 }
+static SSL_CTX* CreateSslCntx() {
```
Comment (adiholden) [inline src/server/replica.cc:143]: why / how is this function different than CreateSslCntx in dragonfly_listerner.cc?

File: src/facade/dragonfly_connection.h [hunk 2]
```diff
@@ -165,6 +165,8 @@ class Connection : public util::Connection {
 
   ConnectionContext* cntx();
 
+  util::FiberSocketBase* parent_;
```
Comment (adiholden) [inline src/facade/dragonfly_connection.h:168]: I believe you dont use this memeber

File: src/facade/dragonfly_listener.h [hunk 3]
```diff
@@ -29,9 +30,11 @@ class Listener : public util::ListenerInterface {
   bool AwaitDispatches(absl::Duration timeout,
                        const std::function<bool(util::Connection*)>& filter);
 
+  SSL_CTX* GetSSLCtx();
```
Comment (adiholden) [inline src/facade/dragonfly_listener.h:33]: I dont see where you use this

File: tests/dragonfly/replication_test.py [hunk 4]
```diff
@@ -1256,3 +1254,29 @@ async def test_no_tls_on_admin_port(df_local_factory, df_seeder_factory, t_maste
     assert 100 == db_size
     await c_replica.close()
     await c_master.close()
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("t_master, t_replica", replication_cases)
+async def test_tls_replication(df_local_factory, df_seeder_factory, t_master, t_replica, with_ca_tls_server_args, with_ca_tls_client_args):
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1260]: can you please add test for reconnect flow

File: src/server/replica.cc (hunk 1 above)
Comment (royjacobson) [reply src/server/replica.cc:143]: Mostly it's creating a client context which is different than a server context, I'd change both functions' names to reflect that.

File: src/server/replica.cc [hunk 5]
```diff
@@ -129,11 +140,49 @@ vector<vector<unsigned>> Partition(unsigned num_flows) {
...
+  SSL_CTX_set_min_proto_version(ctx, TLS1_2_VERSION);
+
+  SSL_CTX_set_options(ctx, SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS);
+
+  SSL_CTX_set_verify(ctx, mask, NULL);
+
+  CHECK_EQ(1, SSL_CTX_set_dh_auto(ctx, 1));
+  return ctx;
+#else
+  return nullptr;
```
Comment (royjacobson) [inline src/server/replica.cc:177]: I'd do

```suggestion
  CHECK(false);
```

File: src/server/replica.cc [hunk 6]
```diff
@@ -129,11 +140,49 @@ vector<vector<unsigned>> Partition(unsigned num_flows) {
...
+#endif
+}
 
 Replica::Replica(string host, uint16_t port, Service* se, std::string_view id)
     : service_(*se), id_{id} {
   master_context_.host = std::move(host);
   master_context_.port = port;
+  ssl_ctx_ = CreateSslCntx();
```
Comment (royjacobson) [inline src/server/replica.cc:185]: I don't think you should create the context unconditionally in the constructor.

File: tests/dragonfly/replication_test.py [hunk 7]
```diff
@@ -1293,3 +1291,57 @@ async def test_no_tls_on_admin_port(
     assert 100 == db_size
     await c_replica.close()
     await c_master.close()
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("t_master, t_replica", replication_cases)
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1297]: please add option of replication over admin port

File: src/server/protocol_client.h [hunk 8]
```diff
@@ -32,6 +36,8 @@ struct JournalReader;
 // This class should be inherited from.
 class ProtocolClient {
  public:
+  using SSL_CTX = struct ssl_ctx_st;
```
Comment (adiholden) [inline src/server/protocol_client.h:39]: will this work if DFLY_USE_SSL if not defined?

File: src/server/protocol_client.cc [hunk 9]
```diff
@@ -39,6 +50,42 @@ using absl::StrCat;
 
 namespace {
 
+#ifdef DFLY_USE_SSL
+static ProtocolClient::SSL_CTX* CreateSslClientCntx() {
+  const bool is_tls_replication_enabled = absl::GetFlag(FLAGS_tls_replication);
+  if (!is_tls_replication_enabled) {
```
Comment (adiholden) [inline src/server/protocol_client.cc:56]: I find this flow a bit confusing.
Why not have this check outside of CreateSslClientCntx and call it only if is_tls_replication_enabled

File: src/server/replica.cc [hunk 10]
```diff
@@ -39,6 +39,8 @@ ABSL_FLAG(int, master_reconnect_timeout_ms, 1000,
           "Timeout for re-establishing connection to a replication master");
 ABSL_DECLARE_FLAG(uint32_t, port);
 
+using namespace std;
```
Comment (adiholden) [inline src/server/replica.cc:42]: you dont have any changes in this file. you can drop this

File: src/server/protocol_client.cc [hunk 11]
```diff
@@ -101,16 +142,49 @@ std::string ProtocolClient::ServerContext::Description() const {
...
+  if (absl::GetFlag(FLAGS_tls_ca_cert_file).empty() &&
+      absl::GetFlag(FLAGS_tls_ca_cert_dir).empty()) {
+    LOG(ERROR) << "Either or both tls_ca_cert_file or tls_ca_cert_dir flags must be set";
+    exit(1);
+  }
+}
+
+void ProtocolClient::MaybeInitSslCtx() {
+  if (absl::GetFlag(FLAGS_tls_replication)) {
+    ValidateFlags();
```
Comment (royjacobson) [inline src/server/protocol_client.cc:160]: A tiny nitpick -
```suggestion
void ProtocolClient::ValidateTlsFlags() const {
  if (absl::GetFlag(FLAGS_tls_cert_file).empty()) {
    LOG(ERROR) << "tls_cert_file flag should be set";
    exit(1);
  }

  if (absl::GetFlag(FLAGS_tls_ca_cert_file).empty() &&
      absl::GetFlag(FLAGS_tls_ca_cert_dir).empty()) {
    LOG(ERROR) << "Either or both tls_ca_cert_file or tls_ca_cert_dir flags must be set";
    exit(1);
  }
}

void ProtocolClient::MaybeInitSslCtx() {
  if (absl::GetFlag(FLAGS_tls_replication)) {
    ValidateTlsFlags();
```

**Review by royjacobson — COMMENTED** [review]
LGTM (small nit)

**Review by adiholden — DISMISSED** [review]

**Review by royjacobson — APPROVED** [review]

**Pull request merged**